sys/include/xtimer.h

```diff
@@ -461,12 +442,30 @@ static inline void xtimer_spin_until(uint32_t value);
 static inline uint32_t xtimer_now(void)
 {
 #if XTIMER_MASK
-    uint32_t a, b;
-    do {
-        a = _lltimer_now() | _high_cnt;
-        b = _lltimer_now() | _high_cnt;
-    } while ((a >> XTIMER_SHIFT_ON_COMPARE) != (b >> XTIMER_SHIFT_ON_COMPARE));
-    return b;
+    typeof(_high_cnt) a, b;
+    typeof(_lltimer_now()) now;
+
+    /* Alternate between latching the high-count and the
+     * low-level timer count. Return when a timer read is
+     * bracketed by two identical high-count reads. */
+
+    a = _high_cnt;
+
+    for (;;) {
+        now = _lltimer_now();
+        b = _high_cnt;
+
+        if (b == a) {
+            return now | b;
+        }
+
+        now = _lltimer_now();
```

Comment:
if you drop this line, and the if (a==b) block below the algorithmic result will be the same but possibly yielding a smaller binary size, and the code becomes easier to read IMO